core/src/modules/ops/destination.ts

```diff
@@ -912,82 +922,115 @@ export namespace DestinationOps {
+        if (gotLock) {
+          locks.push(lock);
+          _exports.push(givenExport);
+        }
       }
 
-      return {
-        success: false,
-        error,
-        retryDelay: outRetryDelay,
-        retryexportIds: _exports.map((e) => e.id),
-      };
-    }
+      if (_exports.length === 0) return;
 
-    let combinedError: CombinedError = null;
-    let exportResult: ExportRecordsPluginMethodResponse;
+      const exportRecords: ExportRecordsPluginMethod = await getBatchFunction(
+        destination
+      );
 
-    try {
-      const exportedRecords = await buildExportedRecords(destination, _exports);
+      const syncMode = await destination.getSyncMode();
+      const syncOperations: DestinationSyncOperations = syncMode
+        ? DestinationSyncModeData[syncMode].operations
+        : DestinationSyncModeData.sync.operations; // if destination does not support sync modes, allow all
 
-      exportResult = await exportRecords({
-        connection,
-        app,
-        appId: app.id,
-        appOptions,
-        destination,
-        destinationId: destination.id,
-        destinationOptions: options,
-        syncOperations,
-        exports: exportedRecords,
+      const options = await destination.getOptions();
+      const app = await destination.$get("app", {
+        scope: null,
+        include: [Option],
       });
-    } catch (error) {
-      combinedError = error;
+      const appOptions = await app.getOptions();
+      const connection = await app.getConnection();
+
+      // check if parallelism ok
+      const parallelismOk = await app.checkAndUpdateParallelism("incr");
+      if (!parallelismOk) {
+        const error = new Error(`parallelism limit reached for ${app.type}`);
+        if (synchronous) throw error;
+
+        const outRetryDelay = config.tasks.timeout + 1;
+        for (const _export of _exports) {
+          await _export.retry(outRetryDelay, true);
+        }
+
+        return {
+          success: false,
+          error,
+          retryDelay: outRetryDelay,
+          retryexportIds: _exports.map((e) => e.id),
+        };
+      }
+
+      let combinedError: CombinedError = null;
+      let exportResult: ExportRecordsPluginMethodResponse;
+
+      try {
+        const exportedRecords = await buildExportedRecords(
+          destination,
+          _exports
+        );
+
+        exportResult = await exportRecords({
+          connection,
+          app,
+          appId: app.id,
+          appOptions,
+          destination,
+          destinationId: destination.id,
+          destinationOptions: options,
+          syncOperations,
+          exports: exportedRecords,
+        });
+      } catch (error) {
+        combinedError = error;
+      } finally {
+        await app.checkAndUpdateParallelism("decr");
+      }
+      return updateExports(
+        destination,
+        _exports,
+        exportResult,
+        combinedError,
+        synchronous
+      );
     } finally {
```

Comment:
Theses nested try/catch/finally areas are a little worrysome to me.  Is there a away that we can only have one try/catch here?  The finally's both are doing cleanup work so it's possible to both release all the locks and fixup the parallelism in one go.  Another variable will be needed to keep track of if we should `app.checkAndUpdateParallelism("decr");`, but I think that's worth if to avoid the nesting

Reply:
The `finally`s are easy to clean up.  It seems like we're depending on things throwing in that "inside" try/catch though ([example](https://github.com/grouparoo/grouparoo/blob/04570d22a1772065d3ca17f288cc8b1904bef7b1/core/__tests__/models/destination/plugins/exportRecords.ts#L866))  I'm sure there's a way to get them unnested.  I have not found it yet though.